### DISCUSSION TITLE
Frage zur Installation

### DISCUSSION THREAD

**Discussion by JAK0721** [flobz/psa_car_controller#447] [General]
Hallo,

ich scheitere an der Installation psa_car_controllers für meinen Corsa e. Ich betreibe einen Headless RPI4
Schon bei 1.1 Download the app on your computer. --> Welche app?
 
1.2 sudo apt-get install python3-typing-extensions... ist ohne Probleme durchgelaufen
1.3 start the app: Start the app with charge control enabled : 
Welche App ist gemeint? Meine Opel App?
Wenn ich den Befehl "psa-car-controller --web-conf" eingebe erhalte ich folgende Meldungen:

pi@raspberrypi:~ $ psa-car-controller --web-conf
Traceback (most recent call last):
  File "/home/pi/.local/bin/psa-car-controller", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/__main__.py", line 19, in main
    app = PSACarController()
  File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/utils/utils.py", line 37, in __call__
    cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
  File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/application/car_controller.py", line 54, in __init__
    self.config = ConfigRepository.read_config()
  File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/repository/config_repository.py", line 164, in read_config
    config.read_string(config_str)
  File "/home/pi/.local/lib/python3.9/site-packages/configupdater/configupdater.py", line 164, in read_string
    return self._parser().read_string(string, source, self)
  File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 312, in read_string
    return self.read_file(sfile, source, into)
  File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 288, in read_file
    self._read(f, source, document)
  File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 495, in _read
    raise MissingSectionHeaderError(fpname, lineno, line)
configparser.MissingSectionHeaderError: File contains no section headers.
file: '<string>', line: 1
'It=1635363162\n'

Ich stehe da echt auf dem Schlauch. Über die Diskussionen und Q&A habe ich auch nicht den passenden Hinweis gefunden.

Danke!

VG, Jörg

**Comment by rimrul** [comment]
> Schon bei 1.1 Download the app on your computer. --> Welche app?

Mit app ist vermutlich einfach der Inhalt der Zipdatei aus dem Release-Bereich gemeint.

>  
> 1.2 sudo apt-get install python3-typing-extensions... ist ohne Probleme durchgelaufen



> 1.3 start the app: Start the app with charge control enabled : 
> Welche App ist gemeint? Meine Opel App?

Nein, `psa-car-controller`. Die Opel-App ist nur unten gemeint wo von `android app` die rede ist.

> Wenn ich den Befehl "psa-car-controller --web-conf" eingebe erhalte ich folgende Meldungen:
> 
> pi@raspberrypi:~ $ psa-car-controller --web-conf
> Traceback (most recent call last):
>   File "/home/pi/.local/bin/psa-car-controller", line 8, in <module>
>     sys.exit(main())
>   File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/__main__.py", line 19, in main
>     app = PSACarController()
>   File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/utils/utils.py", line 37, in __call__
>     cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
>   File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/application/car_controller.py", line 54, in __init__
>     self.config = ConfigRepository.read_config()
>   File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/psacc/repository/config_repository.py", line 164, in read_config
>     config.read_string(config_str)
>   File "/home/pi/.local/lib/python3.9/site-packages/configupdater/configupdater.py", line 164, in read_string
>     return self._parser().read_string(string, source, self)
>   File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 312, in read_string
>     return self.read_file(sfile, source, into)
>   File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 288, in read_file
>     self._read(f, source, document)
>   File "/home/pi/.local/lib/python3.9/site-packages/configupdater/parser.py", line 495, in _read
>     raise MissingSectionHeaderError(fpname, lineno, line)
> configparser.MissingSectionHeaderError: File contains no section headers.
> file: '<string>', line: 1
> 'It=1635363162\n'

Sieht aus als könnte er eine Konfigurationsdatei (config.ini?) Nicht lesen oder als würde da irgendwie Müll drinstehen.

**Reply by JAK0721** [reply]
Danke für Deine Antworten. In der Tat habe ich die Zip Datei nicht runtergeladen. Gibt es einen vordefinierten Zielordner, wo der gesamten Inhalt auf RPI reinkopiert werden sollte?

**Comment by JAK0721** [comment]
oder ist es nicht eher so, dass durch den Befehl "sudo pip3 install psa-car-controller" die Installation automatisch erfolgen sollte?

**Comment by Dirtylife** [comment]
So hab ich´s bei mir hinbekommen:

1. sudo apt-get install python3-...

2. pip3 install psa-car-controller

3. Verzeichnis "psa-car-controller-WorkingDirectory" erstellt

4. in Verzeichnis gewechselt

5. ausführen von "psa-car-controller --web-conf"

6. im Browser unter "http://localhost:5000" konfigurieren

7. systemD-Service eingerichtet und im lokalen LAN verfügbar gemacht

=> zu 6. und 7. kommst Du offenbar gar nicht
     wegen "Error: File contains no section headers"
     
Checke bitte mal Deine "config.ini"
Diese sollte in der ersten Zeile den Header "[General]" enthalten und so aussehen:
(manche Werte habe ich bei mir ergänzt - wichtig ist der Header)

```
[General]
currency = €
# minimum trip length in km so it's added to stats and map in website
minimum trip length = 1
# for future use
length unit = km
[Electricity config]
# price by kw/h
day price = 0.42
night price = 0.42
# ex: 22h30
night hour start = 22h30
# ex: 6h00
night hour end = 06h00
dc charge price =  0.42
high speed dc charge price =  0.42
# minimum power in kW that should be delivered during a charge so it can be considered as a high speed charger
high speed dc charge threshold
charger efficiency = 0.8942
```

**Reply by JAK0721** [reply]
Hallo Dirtylife, danke für Deine Rückmeldung.

Punkt 1 & 2 liefen bereits vorher ohne Problem durch. Punkt 3. bis 5. nach Deiner Anleitung ausgeführt. Hat auch keinen gewünschten Erfolg erzielt (hätte mich ehrlich gesagt auch verwundert). 

Verstehe die ganze Logik der Installation nicht wirklich. Aber liegt wahrscheinlich an meinen mangelnden Linux/Python Kenntnissen.

pi@raspberrypi:~/psa-car-controller-WorkingDirectory $ psa-car-controller --web-conf
2022-12-20 09:15:05,179 :: ERROR :: No config file
2022-12-20 09:15:06,950 :: INFO :: update_data
2022-12-20 09:15:08,243 :: INFO :: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
2022-12-20 09:15:08,244 :: INFO :: Press CTRL+C to quit
^CTraceback (most recent call last):
  File "/home/pi/.local/bin/psa-car-controller", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/psa_car_controller/__main__.py", line 25, in main
    t1.join()
  File "/usr/lib/python3.9/threading.py", line 1033, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.9/threading.py", line 1049, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt

**Reply by JAK0721** [reply]
config.ini wurde mit "psa-car-controller --web-conf" angelegt, allerdings auch mit Fehlermeldungen:
pi@raspberrypi:~/psa-car-controller-WorkingDirectory $ sudo psa-car-controller --web-conf
2022-12-20 09:50:12,287 :: ERROR :: No config file
2022-12-20 09:50:13,727 :: INFO :: update_data
2022-12-20 09:50:13,849 :: INFO :: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
2022-12-20 09:50:13,850 :: INFO :: Press CTRL+C to quit

**Reply by JAK0721** [reply]
Inhalt der config.ini:

```
[General]
currency = €
# minimum trip length in km so it's added to stats and map in website
minimum trip length = 10
# for future use
length unit = km
[Electricity config]
# price by kw/h
day price = 0.15
night price
# ex: 22h30
night hour start
# ex: 6h00
night hour end
dc charge price
high speed dc charge price
# minimum power in kW that should be delivered during a charge so it can be considered as a high speed charger
high speed dc charge threshold
charger efficiency = 0.8942

```

**Comment by Dirtylife** [comment]
OK.

Aber als Du den server per sudo gestartet hast (siehe Dein letzter Post), hats funktioniert...

"Running on http://127.0.0.1:5000/"

Eventuell ein Rechteproblem...

**Reply by JAK0721** [reply]
wenn ich den psa-car-controller wie folgt starte, bekomme ich folgendes zurück

pi@raspberrypi:/usr/local/bin $ sudo psa-car-controller
Traceback (most recent call last):
  File "/usr/local/bin/psa-car-controller", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/__main__.py", line 20, in main
    app.load_app()
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/psacc/application/car_controller.py", line 82, in load_app
    raise FileNotFoundError(self.config_name)
FileNotFoundError: config.json
pi@raspberrypi:/usr/local/bin $

**Reply by JAK0721** [reply]
kann damit allerdings nicht viel anfangen. auf meinem RPI laufen noch der Volkszähler und evcc
Beide Programme rufen ich über die ip des RPI auf, ohne Probleme

**Reply by Dirtylife** [reply]
Dir fehlt die Datei config.json

Diese wird bei der Erstkonfiguration erzeugt.

"sudo psa-car-controller --web-conf" nochmal auführen.
(Hat ja schon mal geklappt)

Danach per Browser auf:
"http://127.0.0.1:5000/" gehen und den Anmeldeprozess durchgehen.

Dann müsste die json-Datei vorhanden sein.

Jetzt kann man den Server entweder manuell oder per Dienst starten. Entweder nur am pi erreichbar oder in LAN oder auch aus den I-net (nicht empfohlen).

**Reply by JAK0721** [reply]
danke für Deine prompte Rückmeldung. 
Leider funktioniert es immer noch nicht. Trotz mehrfacher Ausführung "sudo psa-car-controller --web-conf" fehlt die config.json

**Reply by Dirtylife** [reply]
Und während webconf läuft, hast Du Dich am Browser eingeloggt und die Freischaltung durchlaufen? (Inkl. SMS und allem)

**Reply by JAK0721** [reply]
Hallo Dirtylife, danke für Deine Mühe und vor allem für Deine Geduld.
Ich vermute mein Problem ist, dass ich versuche von meinem Laptop die Einrichtung mit der IP des RPI (192.168.0.240:5000) zu machen.  Besteht auch die Möglichkeit, die Einrichtung vom Laptop zu starten? Mein RPI steckt im Schaltschrank ohne Monitor.

**Reply by Dirtylife** [reply]
OK ok ... Da haben wir nun wohl die Ursache... :-)

http://127.0.0.1:5000/ gilt nur lokal, also wenn Du Dich physisch mit monitor und Tastatur am Pi befindest...

Also, ich fasse mal zusammen:
- Du bist via ssh am Pi eingeloggt
- der Pi hat die Netzwerkadresse  192.168.0.240
- Dein Client hat die Adresse 192.168.0.??? (letzte Zahl ist egal)
- beide sind im selben Netzwerk (Du greifst nicht über das Internet auf den Pi zu)
- "sudo apt-get install python3-..." hast Du am Pi ausgeführt
- "pip3 install psa-car-controller" hast Du am Pi ausgeführt
- Verzeichnis "psa-car-controller-WorkingDirectory" am Pi erstellt
- Du hast in das Verzeichnis gewechselt
- Darin hast Du "sudo psa-car-controller --web-conf" ausgeführt
- und der Server ist erstmal korrekt angelaufen ("Running on http://127.0.0.1:5000/")
=> dadurch ist der Server nur lokal auf der Maschine selbst erreichbar und nicht aus dem lokalen LAN...

Folgender Schritt ist notwendig, um aus dem LAN auf die Website/den Server zugreifen zu können:
"sudo psa-car-controller -l 0.0.0.0 --web-conf"
=> nun auch unter http://192.168.0.240:5000/ erreichbar :-)

- weiter gehts nun am Notebook
- Browser öffnen und Adresse eingeben http://192.168.0.240:5000
- man wird durch die Config geführt - zuerst "User-Config" (kurz warten) und dann "OTP-Config" durchführen
- Nun erzeugt das Programm die config.json
- wenn alles abgeschlossen ist, kann man über ssh den Server beenden "Press CTRL+C to quit"
   (der war bisher ja im Config-Modus...)

Nun ab in den Produktiv-Modus:
 - die Config.ini kann man bei Bedarf an die eigene Werte anpassen, muss man aber nicht
 - zum Testen mal folgendes per ssh im Verzeichnis "psa-car-controller-WorkingDirectory" susführen:
   psa-car-controller -l 0.0.0.0 -f config.json -c charge_config.json -r -d
- Dadurch startet der Server UND ist im LAN erreichbar (Port 5000 sollte aus Sicherheitsgründen nicht aus dem I-Net erreichbar sein)
- Nun zurück in den Browser am Notebook und wieder die Adresse http://192.168.0.240:5000 eingeben
- sollte alles laufen... (Daten sind natürlich noch keine vorhanden, da noch nichts aufgezeichnet)

Falls alles geklappt hat, kann man je nach Gusto auch einen automatischen Start via SystemD anlegen...
(Info: Ich arbeite am Pi mit Raspbian - falls Du nicht, sind die Speicherorte ev abweichend)
(Vorsicht: mein "psa-car-controller-WorkingDirectory" liegt im root-Ordner - anpassen)
(Vorsicht 2: bei mir läufts mit User Root "User=root" - erst mal so versuchen, obs klappt - dann ev auf deinen User umstellen und checken, obs funktioniert) 
- im Verzeichnis "/lib/systemd/system/" die Datei "psa-car-controller.service" erstellen

Inhalt:
```
# add user with this command: adduser --system --no-create-home psa
[Unit]
Description=PSA Controller
After=network-online.target

[Service]
ExecStart=/usr/local/bin/psa-car-controller -l 0.0.0.0 -f config.json -c charge_config.json -r -d
WorkingDirectory=/root/psa-car-controller-WorkingDirectory
Restart=on-failure
RestartSec=1m
Environment=PYTHONUNBUFFERED=true
User=root

[Install]
WantedBy=multi-user.target
```

Den Service kann man nun starten mit: "systemctl start psa-car-controller"
Dann den Status des Service abfragen mit "systemctl status psa-car-controller"
Dann nochmal am Notebook einloggen versuchen
Falls alles geklappt hat, kann man den Service mittels "systemctl enable psa-car-controller" beim Boot aktivieren

**Reply by JAK0721** [reply]
Top - vielen Dank.  Das nenne ich mal eine Anleitung! Hat alles geklappt.
Viele Grüße, Jörg

**Comment by MatzeF1909** [comment]
Moin,
ich bekomme die Installation aufgrund meiner nicht vorhandenen Linux Kenntnisse auch nicht hin.
Ich betreibe ein Rasp Pi 5 mit Raspbian
Ich greife per SSH auf meinen Rasp zu, könnte ihn aber auch an einen Monitor schließen.
Über SSH habe ich den ersten Befehl ("sudo apt-get install python3-..." ) ausgeführt, der auch durchgelaufen ist.
Beim zweiten Befehl (pip3 install psa-car-controller) bekomme ich folgende Fehlermeldung:
pi@raspberrypi:~ $ pip3 install psa-car-controller
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    For more information visit http://rptl.io/venv

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

Ich habe es dann mit dem Befehl "sudo apt install python3-psa-car-controller" was nicht funktioniert hat:
E: Unable to locate package python3-psa-car-controller

Ich stehe total auf dem Schlauch und weiß nicht, wie ich den psa car controller installiert bekomme.
@Dirtylife: Kannst du mir bitte auf die Sprünge helfen?

**Reply by kurim** [reply]
Mach mal folgendes:
python3 -m venv .venv
source .venv/bin/activate
und dann pip3 install psa-car-controller

Dann bist du im [venv](https://docs.python.org/3/library/venv.html) und kannst es installieren

Sollte dann in etwa so aussehen:
![image](https://github.com/flobz/psa_car_controller/assets/1880220/81d40974-6f00-4947-a14d-d1619397d9bb)

**Reply by MatzeF1909** [reply]
Ich habe es jetzt mit dem Zusatz --break-system-packages installiert

Nach der Ausführung von "psa-car-controller -l 0.0.0.0 --web-conf" bekomme ich jetzt folgende Meldung:
pi@raspberrypi:~ $ psa-car-controller -l 0.0.0.0 --web-conf
2024-02-25 17:01:30,454 :: INFO :: App version 3.4.0
2024-02-25 17:01:30,454 :: ERROR :: No config file
2024-02-25 17:01:30,568 :: INFO :: NumExpr defaulting to 4 threads.
2024-02-25 17:01:31,021 :: INFO :: update_data
2024-02-25 17:01:31,046 :: INFO :: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5000
 * Running on http://192.168.178.115:5000
2024-02-25 17:01:31,046 :: INFO :: Press CTRL+C to quit
^CTraceback (most recent call last):
  File "/home/pi/.local/bin/psa-car-controller", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/pi/.local/lib/python3.11/site-packages/psa_car_controller/__main__.py", line 25, in main
    t1.join()
  File "/usr/lib/python3.11/threading.py", line 1112, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.11/threading.py", line 1132, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Wenn ich 192.168.178.115:5000 kommt leider nichts.

Soll ich "deinen Weg" noch einmal versuchen oder woran liegt es?

**Reply by kurim** [reply]
Ja das mit " --break-system-packages" hatte ich auch gelesen halte ich nur für eine unschöne Lösung.

**Comment by MatzeF1909** [comment]
Wenn ich deinen Befehl eingebe, bekomme ich das hier raus:
pi@raspberrypi:~ $ python3 -m venv .venv
Error: Command '['/home/pi/.venv/bin/python3', '-m', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.

**Reply by kurim** [reply]
Hm, gute frage...
python3 -V => 3.11 ?

ggf. die venv noch mal löschen:
rm -rf .venv

Unterumständen musst du pip noch mal updaten:
python3 -m pip install --upgrade pip

**Reply by MatzeF1909** [reply]
Dann bekomme ich wieder das:
pi@raspberrypi:~ $ python3 -m pip install --upgrade pip
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    For more information visit http://rptl.io/venv

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

**Comment by MatzeF1909** [comment]
Nach einem Reboot des Rasp + Ausführung der "psa-car-controller -l 0.0.0.0 --web-conf" hat sich das Webinterface nun geöffnet
Mühsam ernährt sich....
Ich versuche es jetzt mal weiter und hoffe, dass ich es hinbekomme

**Comment by MatzeF1909** [comment]
Wenn ich meine Zugangsdaten eingebe, steht auf der Webseite folgender Fehler:
[Errno 28] No space left on device


Ich habe "psa-car-controller -l 0.0.0.0 --web-conf" aus dem Verzeichnis /home/pi/psa-car-controller-WorkingDirectory gestartet, dass ich zuvor angelegt und in das ich die Dateien und Verzeichnisse aus der runtergeladenen .zip kopiert habe.

Die SD Karte ist aber nicht voll...da ist nur das Raspbian OS, evcc und iobroker mit ca. 10 Adaptern drauf bei einer 64GB Karte
Was nun?

**Comment by MatzeF1909** [comment]
pi@raspberrypi:~/psa-car-controller-WorkingDirectory $ df
Filesystem                                        1K-blocks       Used Available Use% Mounted on
udev                                                3949056          0   3949056   0% /dev
tmpfs                                                824704       5680    819024   1% /run
/dev/mmcblk0p2                                      8181612    7780744         0 100% /
tmpfs                                               4123504        288   4123216   1% /dev/shm
tmpfs                                                  5120         48      5072   1% /run/lock
/dev/mmcblk0p1                                       522232      74164    448068  15% /boot/firmware
tmpfs                                                824688        144    824544   1% /run/user/1000

**Comment by MatzeF1909** [comment]
Scheinbar habe ich mir mit "--break-system-packages" nun mein System gehimmelt...es ist voll
Ich setze dann mal alles neu auf und nehme einen neuen Anlauf

**Comment by MatzeF1909** [comment]
Also...System neu gemacht
Dateien in ein neues Verzeichnis psa/psa_car_controller-3.4.0 kopiert
in das Verzeichnis navigiert
python3 -m venv .venv
source .venv/bin/activate
pip3 install psa-car-controller
psa-car-controller -l 0.0.0.0 --web-conf
Zugang hergestellt
psa-car-controller -l 0.0.0.0 -f config.json -c charge_config.json -r -d

Was soll ich sagen...es scheint zu laufen.

Jetzt muss ich noch so einstellen, dass es immer läuft und automatisch startet
Wie mache ich das, wenn es im venv ist?
Datei habe ich schon im Verzeichnis /lib/systemd/system/ angelegt:

[Unit]
Description=PSA Controller
After=network-online.target

[Service]
ExecStart=/usr/local/bin/psa-car-controller -l 0.0.0.0 -f config.json -c charge_config.json -r -d
WorkingDirectory=/home/pi/psa/psa_car_controller-3.4.0
Restart=on-failure
RestartSec=1m
Environment=PYTHONUNBUFFERED=true
User=root

[Install]
WantedBy=multi-user.target

Wenn ich den Service starten will kommt:
pi@raspberrypi:/lib/systemd/system $ sudo systemctl start psa-car-controller
Failed to start psa-car-controller.service: Unit psa-car-controller.service is masked.

Ich vermute bei "WorkingDirectory" muss etwas anderes rein...richtig, damit der Befehl aus dem venv gestartet wird, oder?

**Reply by kurim** [reply]
evtl hilft das:
https://stackoverflow.com/questions/37211115/how-to-enable-a-virtualenv-in-a-systemd-service-unit

**Reply by MatzeF1909** [reply]
Ich bekomme das nicht hin
Gibt es nicht auch die Möglichkeit ein Script zu schreiben, dass die Befehle ausführt...
cd psa
cd psa_car_controller-3.4.1
source .venv/bin/activate
psa-car-controller -l 0.0.0.0 -f config.json -c charge_config.json -r -d

Das Script dann in die Autostart und schon sollte es doch laufen, oder denke ich zu einfach?

Man kann dann leider nur den Service nicht einfach stoppen, schätze ich

**Reply by hijacker123** [reply]
@MatzeF1909

Ich stand vor dem gleichen Problem, einer virtuellen python-Instanz bzw. Installation. Der "Trick" ist einfach "ignorant" den psa-car-controller innerhalb der virtuellen Umgebung bzw. dem Pfad aufzurufen.

Folgende Infos benötigst Du:
1.) Deinen Username, bei mir ist es **rock**
2.) Dein Home-Verzeichnis, bei mir ist es **/home/rock** (dort liegen auch die Config-Dateien)
3.) Den Pfad Deiner erstellten virtuellen python-Umgebung in dem nun das `psa-car-controller` liegt, bei mir ist es **/home/rock/venv-test-psa/bin**
4.) Die IP-Adresse unter der Deine Installation/Dein Rechner erreichbar ist, bei mir ist es **192.168.180.153**

Mit diesen Infos erstellst Du den psa-car-controller.service und ersetzt die Angaben für User/Pfad/IP durch die Daten von Deiner Installation

`sudo nano /etc/systemd/system/psa-car-controller.service`

```
[Unit]
Description=psa-car-controller
After=syslog.target network.target

[Service]
Type=simple
User=rock
WorkingDirectory=/home/rock
ExecStart=/home/rock/venv-test-psa/bin/psa-car-controller -l 192.168.180.153 -f /home/rock/config.json -c /home/rock/charge_config.json -r -d 
Restart=on-abort

[Install]
WantedBy=multi-user.target
```

Danach einfach testen

ggfs. `sudo systemctl enable psa-car-controller`
`sudo systemctl start psa-car-controller`
`sudo systemctl status psa-car-controller` und auf Fehler bei der Konsolenausgabe achten. Prüfe die Erreichbarkeit im Browser.

**Comment by alexandererny** [comment]
Hallo zusammen,
ich brauche mal eure Hilfe bei der Installation auf meinem Raspi Zero. Den Raspi kontrolliere ich per ssh von meinem Laptop aus. So habe ich nun schon Python und den psa-car-controller installiert, und kann den psa-car-controller auch lokal starten. Da ich aber keinen lokalen Zugang habe, möchte ich die Web Confirguration aus dem LAN machen. Ich schaffe es jedoch nicht, den psa-car controller mit 
 sudo psa-car-controller -l 0.0.0.0 --web-conf
zu starten. Ich erhalte lediglich die Meldung 
sudo: psa-car-controller: command not found
Habt ihr eine Idee wie ich heir weiterkomme?

Danke & Grüße

**Reply by Sillium007** [reply]
Kommt halt drauf an wie Du es installiert hast. Ich vermute Du hast es für den Nutzer (anscheinen nicht "root") lokal installiert. Anscheinend geht es ja auch es "lokal" zu starten wie Du schreibst.
Wenn Du es jetzt mit einem anderen Nutzer (root) versuchst liegt das Binary nicht im Pfad und kann daher nicht gefunden werden.

Ich würde spontan mal mit den Nutzer mit dem es geht "which psa-car-controller" machen dann hast Du die genaue Stelle wo Du sie hin installiert hast.

**Reply by alexandererny** [reply]
Danke für die Hilfe, werd's so versuchen!

**Comment by ich77** [comment]
Hallo,
ich scheitere leider aufgrund meiner nur rudimentär vorhandenden Linuxkenntnisse an der Installation. Ich weiß nicht, welche ZIP Datei soll ich von wo und wie auf den Raspberry kopieren? 
Danke
Markus

**Reply by Sillium007** [reply]
Ein Zip eines der Releases: https://github.com/flobz/psa_car_controller/releases

**Comment by Easyrunner** [comment]
@hijacker123 
Leider schaffe ich es auch nicht PSA-CC automatisch SystemD service in einer virtuellen Umgebung automatisch zu starten.
Hier meine psa-car-controller.service Datei

`[Unit]
Description=psa-car-controller
After=syslog.target network.target

[Service]
Type=simple
User=pi
WorkingDirectory=/home/pi
ExecStart=/home/pi/.venv/bin/psa-car-controller -f config.json -c charge_config.json -r -l 192.168.178.37

Restart=on-abort

[Install]
WantedBy=multi-user.target

`

Den venv-Pfad habe ich direkt so rauskopiert und stimmt, Die IP Adresse von dem Pi worauf es läuft auch.

Auch ...
`ExecStart=/home/rock/venv-test-psa/bin/psa-car-controller -l 192.168.180.153 -f /home/rock/config.json -c /home/rock/charge_config.json -r -d `
... mit geänderten Pfaden und IP Adresse habe ich probiert.


Der Status meldet immer:

(.venv) pi@RasPi4:~ $ sudo systemctl status psa-car-controller
× psa-car-controller.service - psa-car-controller
     Loaded: loaded (/etc/systemd/system/psa-car-controller.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Thu 2025-02-27 18:16:38 CET; 14h ago
   Duration: 12.064s
    Process: 657 ExecStart=/home/pi/.venv/bin/psa-car-controller -f config.json -c charge_config.json -r -l 192.168.178.37 (code=exited, status=1/FAILURE)
   Main PID: 657 (code=exited, status=1/FAILURE)
        CPU: 6.020s

Feb 27 18:16:37 RasPi4 psa-car-controller[657]:              ^^^^^^
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:   File "/home/pi/.venv/lib/python3.11/site-packages/psa_car_controller/__main__.py", line 20, in main
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:     app.load_app()
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:   File "/home/pi/.venv/lib/python3.11/site-packages/psa_car_controller/psacc/application/car_controller.py", lin>
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:     raise ConnectionError
Feb 27 18:16:37 RasPi4 psa-car-controller[657]: ConnectionError
Feb 27 18:16:37 RasPi4 psa-car-controller[657]: 2025-02-27 18:16:37,920 :: INFO :: save config change
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Main process exited, code=exited, status=1/FAILURE
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Failed with result 'exit-code'.
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Consumed 6.020s CPU time.
lines 1-18/18 (END)...skipping...
× psa-car-controller.service - psa-car-controller
     Loaded: loaded (/etc/systemd/system/psa-car-controller.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Thu 2025-02-27 18:16:38 CET; 14h ago
   Duration: 12.064s
    Process: 657 ExecStart=/home/pi/.venv/bin/psa-car-controller -f config.json -c charge_config.json -r -l 192.168.178.37 (code=exited, status=1/FAILURE)
   Main PID: 657 (code=exited, status=1/FAILURE)
        CPU: 6.020s

Feb 27 18:16:37 RasPi4 psa-car-controller[657]:              ^^^^^^
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:   File "/home/pi/.venv/lib/python3.11/site-packages/psa_car_controller/__main__.py", line 20, in main
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:     app.load_app()
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:   File "/home/pi/.venv/lib/python3.11/site-packages/psa_car_controller/psacc/application/car_controller.py", line 118, in load_app
Feb 27 18:16:37 RasPi4 psa-car-controller[657]:     raise ConnectionError
Feb 27 18:16:37 RasPi4 psa-car-controller[657]: ConnectionError
Feb 27 18:16:37 RasPi4 psa-car-controller[657]: 2025-02-27 18:16:37,920 :: INFO :: save config change
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Main process exited, code=exited, status=1/FAILURE
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Failed with result 'exit-code'.
Feb 27 18:16:38 RasPi4 systemd[1]: psa-car-controller.service: Consumed 6.020s CPU time.


Eine Idee?

**Comment by Easyrunner** [comment]
Ich musste noch einmal den Authentifikationsprozess durchlaufen, jetzt geht es wieder auch über SystemD

Wie oft muss man diesen Prozess neu durchlaufen?
